File: app/src/androidTest/java/br/com/jonathanzanella/myexpenses/resume/CalculateMonthBalanceCorrectlyTest.java
package br.com.jonathanzanella.myexpenses.resume;

import android.content.Intent;
import android.support.test.InstrumentationRegistry;
import android.support.test.filters.SmallTest;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;

import org.joda.time.DateTime;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.account.Account;
import br.com.jonathanzanella.myexpenses.account.AccountRepository;
import br.com.jonathanzanella.myexpenses.bill.Bill;
import br.com.jonathanzanella.myexpenses.bill.BillRepository;
import br.com.jonathanzanella.myexpenses.database.DatabaseHelper;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.expense.Expense;
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository;
import br.com.jonathanzanella.myexpenses.helpers.ActivityLifecycleHelper;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import br.com.jonathanzanella.myexpenses.helpers.builder.AccountBuilder;
import br.com.jonathanzanella.myexpenses.helpers.builder.BillBuilder;
import br.com.jonathanzanella.myexpenses.helpers.builder.ExpenseBuilder;
import br.com.jonathanzanella.myexpenses.helpers.builder.ReceiptBuilder;
import br.com.jonathanzanella.myexpenses.helpers.builder.SourceBuilder;
import br.com.jonathanzanella.myexpenses.receipt.Receipt;
import br.com.jonathanzanella.myexpenses.receipt.ReceiptRepository;
import br.com.jonathanzanella.myexpenses.source.Source;
import br.com.jonathanzanella.myexpenses.source.SourceRepository;
import br.com.jonathanzanella.myexpenses.views.MainActivity;

import static android.support.test.InstrumentationRegistry.getInstrumentation;
import static android.support.test.InstrumentationRegistry.getTargetContext;
import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDescendantOfA;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static br.com.jonathanzanella.myexpenses.helpers.UIHelper.clickIntoView;
import static junit.framework.Assert.assertTrue;
import static org.hamcrest.Matchers.allOf;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class CalculateMonthBalanceCorrectlyTest {
	private static final int EXPENSE_VALUE = 100;
	private static final int RECEIPT_INCOME = 200;
	private static final int BILL_AMOUNT = 25;

	@Rule
	public ActivityTestRule<MainActivity> activityTestRule = new ActivityTestRule<>(MainActivity.class);

	private final MonthlyPagerAdapterHelper monthlyPagerAdapterHelper = new MonthlyPagerAdapterHelper();

	@Before
	public void setUp() throws Exception {
		new DatabaseHelper(InstrumentationRegistry.getTargetContext()).recreateTables();

		Account a = new AccountBuilder().build();
		assertTrue(new AccountRepository().save(a).isValid());

		Source s = new SourceBuilder().build();
		assertTrue(new SourceRepository().save(s).isValid());

		DateTime now = DateTime.now().withDayOfMonth(1);
		Bill b = new BillBuilder()
				.initDate(now)
				.endDate(now.plusMonths(12))
				.amount(BILL_AMOUNT)
				.build();
		assertTrue(new BillRepository(new ExpenseRepository(new RepositoryImpl<>(getTargetContext()))).save(b).isValid());

		generateThreeMonthlyReceipts(a, s);
		generateThreeMonthlyExpenses(a);
	}

	@After
	public void tearDown() throws Exception {
		ActivityLifecycleHelper.closeAllActivities(getInstrumentation());
	}

	private void generateThreeMonthlyReceipts(Account a, Source s) {
		DateTime dateTime = DateTime.now();
		ReceiptRepository receiptRepository = new ReceiptRepository(new RepositoryImpl<>(getTargetContext()));
		Receipt r = new ReceiptBuilder()
				.account(a)
				.source(s)
				.date(dateTime)
				.income(RECEIPT_INCOME)
				.build();
		assertTrue(receiptRepository.save(r).isValid());
		dateTime = dateTime.plusMonths(1);
		r = new ReceiptBuilder()
				.account(a)
				.source(s)
				.date(dateTime)
				.income(RECEIPT_INCOME * 2)
				.build();
		assertTrue(receiptRepository.save(r).isValid());
		dateTime = dateTime.plusMonths(1);
		r = new ReceiptBuilder()
				.account(a)
				.source(s)
				.date(dateTime)
				.income(RECEIPT_INCOME * 3)
				.build();
		assertTrue(receiptRepository.save(r).isValid());
	}

	private void generateThreeMonthlyExpenses(Account a) {
		DateTime dateTime = DateTime.now();
		ExpenseRepository expenseRepository = new ExpenseRepository(new RepositoryImpl<>(getTargetContext()));
		Expense r = new ExpenseBuilder()
				.chargeable(a)
				.date(dateTime)
				.value(EXPENSE_VALUE)
				.build();
		assertTrue(expenseRepository.save(r).isValid());
		dateTime = dateTime.plusMonths(1);
		r = new ExpenseBuilder()
				.chargeable(a)
				.date(dateTime)
				.value(EXPENSE_VALUE * 2)
				.build();
		assertTrue(expenseRepository.save(r).isValid());
		dateTime = dateTime.plusMonths(1);
		r = new ExpenseBuilder()
				.chargeable(a)
				.date(dateTime)
				.value(EXPENSE_VALUE * 3)
				.build();
		assertTrue(expenseRepository.save(r).isValid());
	}

	@Test
	public void verify_month_balance() throws Exception {
		activityTestRule.launchActivity(new Intent());

		int balance = RECEIPT_INCOME - EXPENSE_VALUE - BILL_AMOUNT;
		int twoMonthsBalance = RECEIPT_INCOME * 3 - EXPENSE_VALUE * 3 - BILL_AMOUNT;
		String expectedBalance = CurrencyHelper.INSTANCE.format(balance);
		String twoMonthsExpectedBalance = CurrencyHelper.INSTANCE.format(twoMonthsBalance);

		validateExpectedBalance(expectedBalance);

		scrollToMonth(DateTime.now().plusMonths(2));
		validateExpectedBalance(twoMonthsExpectedBalance);

		scrollToMonth(DateTime.now());
		validateExpectedBalance(expectedBalance);
	}

	private void validateExpectedBalance(String expectedBalance) {
		onView(allOf(
				withId(R.id.balance),
				isDescendantOfA(allOf(
						withId(R.id.view_monthly_resume),
						isDisplayed()))))
				.check(matches(withText(expectedBalance)));
	}

	private void scrollToMonth(DateTime dateTime) {
		String twoMonthsFromNow = monthlyPagerAdapterHelper.formatMonthForView(dateTime);
		clickIntoView(twoMonthsFromNow);
	}
}

File: app/src/androidTest/java/br/com/jonathanzanella/myexpenses/resume/CalculateMonthBalanceCorrectlyTest.kt
package br.com.jonathanzanella.myexpenses.resume

import android.content.Intent
import android.support.test.InstrumentationRegistry.getInstrumentation
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.assertion.ViewAssertions.matches
import android.support.test.espresso.matcher.ViewMatchers
import android.support.test.filters.SmallTest
import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import android.view.View
import br.com.jonathanzanella.TestApp
import br.com.jonathanzanella.myexpenses.App
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.account.AccountDataSource
import br.com.jonathanzanella.myexpenses.bill.BillDataSource
import br.com.jonathanzanella.myexpenses.expense.ExpenseDataSource
import br.com.jonathanzanella.myexpenses.helpers.ActivityLifecycleHelper
import br.com.jonathanzanella.myexpenses.helpers.UIHelper.clickIntoView
import br.com.jonathanzanella.myexpenses.helpers.builder.BillBuilder
import br.com.jonathanzanella.myexpenses.helpers.builder.ExpenseBuilder
import br.com.jonathanzanella.myexpenses.helpers.builder.ReceiptBuilder
import br.com.jonathanzanella.myexpenses.helpers.builder.SourceBuilder
import br.com.jonathanzanella.myexpenses.helpers.toCurrencyFormatted
import br.com.jonathanzanella.myexpenses.receipt.ReceiptDataSource
import br.com.jonathanzanella.myexpenses.source.Source
import br.com.jonathanzanella.myexpenses.source.SourceRepository
import br.com.jonathanzanella.myexpenses.views.MainActivity
import junit.framework.Assert.assertTrue
import org.hamcrest.Matchers.allOf
import org.joda.time.DateTime
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import javax.inject.Inject

@RunWith(AndroidJUnit4::class)
@SmallTest
class CalculateMonthBalanceCorrectlyTest {

    @Rule @JvmField
    var activityTestRule = ActivityTestRule(MainActivity::class.java)

    @Inject
    lateinit var billDataSource: BillDataSource
    @Inject
    lateinit var sourceRepository: SourceRepository
    @Inject
    lateinit var accountDataSource: AccountDataSource
    @Inject
    lateinit var receiptDataSource: ReceiptDataSource
    @Inject
    lateinit var expenseDataSource: ExpenseDataSource

    private val monthlyPagerAdapterHelper = MonthlyPagerAdapterHelper()

    @Before
    @Throws(Exception::class)
    fun setUp() {
        TestApp.getTestComponent().inject(this)
        App.resetDatabase()

        val a = br.com.jonathanzanella.myexpenses.helpers.builder.AccountBuilder().build()
        assertTrue(accountDataSource.save(a).isValid)

        val s = SourceBuilder().build()
        assertTrue(sourceRepository.save(s).isValid)

        val now = DateTime.now().withDayOfMonth(1)
        val b = BillBuilder()
                .initDate(now)
                .endDate(now.plusMonths(12))
                .amount(BILL_AMOUNT)
                .build()
        assertTrue(billDataSource.save(b).isValid)

        generateThreeMonthlyReceipts(a, s)
        generateThreeMonthlyExpenses(a)
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        ActivityLifecycleHelper.closeAllActivities(getInstrumentation())
    }

    private fun generateThreeMonthlyReceipts(a: Account, s: Source) {
        var dateTime = DateTime.now()
        var r = ReceiptBuilder()
                .account(a)
                .source(s)
                .date(dateTime)
                .income(RECEIPT_INCOME)
                .build()
        assertTrue(receiptDataSource.save(r).isValid)
        dateTime = dateTime.plusMonths(1)
        r = ReceiptBuilder()
                .account(a)
                .source(s)
                .date(dateTime)
                .income(RECEIPT_INCOME * 2)
                .build()
        assertTrue(receiptDataSource.save(r).isValid)
        dateTime = dateTime.plusMonths(1)
        r = ReceiptBuilder()
                .account(a)
                .source(s)
                .date(dateTime)
                .income(RECEIPT_INCOME * 3)
                .build()
        assertTrue(receiptDataSource.save(r).isValid)
    }

    private fun generateThreeMonthlyExpenses(a: Account) {
        var dateTime = DateTime.now()
        var r = ExpenseBuilder()
                .chargeable(a)
                .date(dateTime)
                .value(EXPENSE_VALUE)
                .build()
        assertTrue(expenseDataSource.save(r).isValid)
        dateTime = dateTime.plusMonths(1)
        r = ExpenseBuilder()
                .chargeable(a)
                .date(dateTime)
                .value(EXPENSE_VALUE * 2)
                .build()
        assertTrue(expenseDataSource.save(r).isValid)
        dateTime = dateTime.plusMonths(1)
        r = ExpenseBuilder()
                .chargeable(a)
                .date(dateTime)
                .value(EXPENSE_VALUE * 3)
                .build()
        assertTrue(expenseDataSource.save(r).isValid)
    }

    @Test
    @Throws(Exception::class)
    fun verify_month_balance() {
        activityTestRule.launchActivity(Intent())

        val balance = RECEIPT_INCOME - EXPENSE_VALUE - BILL_AMOUNT
        val twoMonthsBalance = RECEIPT_INCOME * 3 - EXPENSE_VALUE * 3 - BILL_AMOUNT
        val expectedBalance = balance.toCurrencyFormatted()
        val twoMonthsExpectedBalance = twoMonthsBalance.toCurrencyFormatted()

        validateExpectedBalance(expectedBalance)

        scrollToMonth(DateTime.now().plusMonths(2))
        validateExpectedBalance(twoMonthsExpectedBalance)

        scrollToMonth(DateTime.now())
        validateExpectedBalance(expectedBalance)
    }

    private fun validateExpectedBalance(expectedBalance: String) {
        onView(allOf<View>(
                ViewMatchers.withId(R.id.balance),
                ViewMatchers.isDescendantOfA(allOf<View>(
                        ViewMatchers.withId(R.id.view_monthly_resume),
                        ViewMatchers.isDisplayed()))))
                .check(matches(ViewMatchers.withText(expectedBalance)))
    }

    private fun scrollToMonth(dateTime: DateTime) {
        val twoMonthsFromNow = monthlyPagerAdapterHelper.formatMonthForView(dateTime)
        clickIntoView(twoMonthsFromNow)
    }

    companion object {
        private val EXPENSE_VALUE = 100
        private val RECEIPT_INCOME = 200
        private val BILL_AMOUNT = 25
    }
}
